### PR TITLE
feat: add defer_save, cancel_save, immediate_save trigger options

### LINKED ISSUE #8: Save only after sequence of actions finished, not after each action
Copied from https://github.com/Pocco81/auto-save.nvim/issues/72

## Problem:
When I edit text, I do it in a `sequence` of  multiple `actions`, that I execute rapidly after each other.
For example, I delete a word (`daw`), change a word (`caw`, after that typing and exiting insert mode), insert the deleted word two lines below.
With the current implementation that would save up to 4 times (`TextChanged`, `TextChanged`, `InsertLeave`, `TextChanged`), depending on my typing speed and `debounce_delay`.
This triggers any external service like Webpack up to 4 times, once per `action`.

I want it to be only triggered 1 time, once per `sequence`.

## Idea:
Add 3 types of events: `ImmediateSave`, `DeferSave` and `CancelDeferedSave` 

- when `DeferSave` event (`TextChanged`, `InsertLeave`, ...) happens, start a timer `t`, which will save the file if it finishes
- when `CancelDeferedSave` event (`InsertEnter`, `TextChanged`, enter, ...) happens, cancel the timer for this file
- when `ImmediateSave` events (BufferLeave, VimClose, ...) happen, save immediately

This makes it so, that `actions` that occur fast after another will cancel each other out and only the last action gets saved.
The Timer would have to be something like 2 seconds.

Sometimes however we need to save immediately, for example when closing/leaving a buffer or closing nvim.
So we need another set of events called `ImmediateSave`.

### PR DESCRIPTION
**Pull request by primeapple**
This fixes #8

### CODE REVIEW

**Comment by primeapple** [comment]
Just from the codestructure I'm happy now.
I'll try using it from next week on and see if there are any bugs.

@okuuva feel free to have a look over it.

I also added some `TODO`s that might be good for later improvements.

File: lua/auto-save/init.lua [hunk 1]
```diff
@@ -98,23 +114,49 @@ function M.save(buf)
...
+    save_func = (cnf.opts.debounce_delay > 0 and debounce(save, cnf.opts.debounce_delay) or save)
   end
   save_func()
 end
 
 function M.on()
-  api.nvim_create_autocmd(cnf.opts.trigger_events, {
+  api.nvim_create_autocmd(cnf.opts.trigger_events.immediate_save, {
+    callback = M.immediate_save,
+    pattern = "*",
```
Comment (okuuva) [inline lua/auto-save/init.lua:141]: Isn't `pattern = "*",` the default?

**Comment by okuuva** [comment]
Excellent work @primeapple! I'll take it into daily use as well to spot any potential problems but LGTM already :)

File: README.md [hunk 2]
```diff
@@ -122,7 +126,7 @@ EOF
     return false -- can't save
   end,
   write_all_buffers = false, -- write all buffers when the current one meets `condition`
-  debounce_delay = 135, -- saves the file at most every `debounce_delay` milliseconds
+  debounce_delay = 135, -- delay after which a pending save is executed
```
Comment (okuuva) [inline README.md:129]: To me this sounds a very short default value now with the `defer_save` and `cancel_save`. I'd go as far as set it two something like 2000ms but I'll do that first on my own config and I'll report back on how it works.

File: lua/auto-save/config.lua [hunk 3]
```diff
@@ -9,7 +9,11 @@ Config = {
       dim = 0.18, -- dim the color of `message`
       cleaning_interval = 1250, -- (milliseconds) automatically clean MsgArea after displaying `message`. See :h MsgArea
     },
-    trigger_events = { "InsertLeave", "TextChanged" }, -- vim events that trigger auto-save. See :h events
+    trigger_events = { -- See :h events
+      immediate_save = { "BufLeave", "FocusLost" }, -- vim events that trigger an immediate save
```
Comment (okuuva) [inline lua/auto-save/config.lua:13]: Not sure about "FocusLost" here but then again don't see any hard reason not to save immediately then so maybe that's just me.

**Comment by okuuva** [comment]
And I do agree with the TODOs. I wouldn't maybe drop those in this MR just yet just to keep things as atomic as possible. And I need to comb through the code a few more times since I've also wondered what's the point with those seemingly unused things. Probably remnants from the rewrite Pocco81 did in July 2022 but need to triple check.

File: README.md (hunk 2 above)
Comment (primeapple) [reply README.md:129]: Totally true! I actually have set it to 2000ms on my local config as well.
I'll toy around with it a bit, I don't see why we shouldn't go with 2000ms.

File: lua/auto-save/config.lua (hunk 3 above)
Comment (primeapple) [reply lua/auto-save/config.lua:13]: What's your reason against `FocusLost`? I have some obscure workflows (editing with `nvim` in tab 1, then switching to tab 2 and running the tests from there). This was the reason for including this :)

Ultimately we should maybe add a section to the README, that includes information for useful events.

File: lua/auto-save/init.lua (hunk 1 above)
Comment (primeapple) [reply lua/auto-save/init.lua:141]: Where do I get information about the defaults for the `nvim` api? I didn't see them in here: https://neovim.io/doc/user/api.html :(

But I think you are right with this!

**Comment by primeapple** [comment]
> And I do agree with the TODOs. I wouldn't maybe drop those in this MR just yet just to keep things as atomic as possible.

I'll remove the todos and open another PR afterwards to add them, maybe even removing the potential unused code.

**Comment by primeapple** [comment]
I have some trouble because it doesn't seem to be possible to set timers to buffer variables, see here: https://www.reddit.com/r/neovim/comments/123y048/cannot_convert_userdata_when_trying_to_set_a/

I might just use the `global_vars` variable after all...

File: README.md (hunk 2 above)
Comment (primeapple) [reply README.md]: I went with 1000ms for now :)

File: lua/auto-save/init.lua (hunk 1 above)
Comment (primeapple) [reply lua/auto-save/init.lua]: removed all `pattern = "*"` lines

**Comment by primeapple** [comment]
Ok. I open this as "Ready for review". It works quite well on my machine now, you are save to try it out @okuuva :)

**Review by okuuva — APPROVED** [review]
Haven't had time to test this thoroughly since I was away for the weekend and been sick since I got back. But couldn't find any obvious problems with the short time I had to test it, the code looks good to me and I trust you've been testing it yourself during the development much more rigorously than I have. Exceptional job!

**Comment by primeapple** [comment]
Ok, thank you!
Let's observe how it behaves in the future ;-)

**Pull request merged**